### ISSUE TITLE
OpenAI-compatible provider: `repetitionPenalty` setting has no effect (wire-level fix needed)

### ISSUE THREAD

**Issue by green-dalii** [green-dalii/obsidian-llm-wiki#414]
## Background

Split from #402 (now closed). `response_format` was the easy part — `ca4a24d` (eucher, 2026-07-29) removed it from the wire path because LM Studio rejects `{"type":"json_object"}` outright (`response_format.type` must be `json_schema` or `text` per OpenAI's Chat Completions spec, which LM Studio implements). That removal changes nothing on the wire because under the wrong `providerOptions.openaiCompatible` key the field never reached the server anyway.

## What remains

`repetitionPenalty` (a.k.a. the #128 user setting) is still emitted at `src/llm-sdk/openai-compat-sdk-client.ts:417`:

```ts
openaiOpts.repetition_penalty = opts.repetitionPenalty;
```

under the `"openaiCompatible"` key in the returned `providerOptions`. The SDK's zod schema (`@ai-sdk/openai-compatible@2.0.62/dist/index.mjs:322-344`) only knows `reasoningEffort`, `textVerbosity`, `strictJsonSchema`, and `user` for chat options — everything else is filtered out. Net effect: every LM Studio / llama.cpp / vLLM backend silently ignores the `repetitionPenalty` user setting.

## Why it's harder than `response_format`

Two viable paths, neither trivially safe:

1. **`transformRequestBody` route** — same mechanism used for the `max_tokens` → `max_completion_tokens` swap at line 114-127. DocTpoint verified (Issue #402 comment 1, 2026-08-03 13:35 UTC) that routing `response_format` through `transformRequestBody` breaks LM Studio. `repetition_penalty` is a different field with different backend compatibility — needs a fresh measurement, not a parallel assumption.

2. **Switch the wrapper key** to `this.providerOptionsName` (i.e. `this.provider`, the actual provider id — `deepseek` / `kimi` / `lmstudio` / `custom` / `ollama`). This makes the SDK's `filter()` passthrough path (line 533-540) actually find the fields. **Caveat**: the same passthrough path that would resurrect `thinking` and `chat_template_kwargs` — fields we explicitly want stripped for #411's force-disable-thinking fix. Per-id key correction is on the v1.26.x PATCH track ([[project_v1_26_0_rescoped]]) as a separate dependency.

## Why I'm not opening a fix PR immediately

(1) needs per-backend compatibility measurement (LM Studio / Ollama / DeepSeek / Kimi / GLM each handle `repetition_penalty` differently). (2) collides with the per-id key correction deferred from v1.26.0 Batch 6 — would re-introduce the `thinking` / `chat_template_kwargs` strip-back if the v1.26.x PATCH fix is wrong. Both paths need design review before code.

## Acceptance criteria for the fix

- `repetitionPenalty: 1.1` produces measurably different output vs `repetitionPenalty: 1.0` on a real LM Studio / gemma-4-12b run (output token repetition rate at the prompt tail is the standard signal)
- The fix does not regress #411's force-disable thinking (reasoning tokens remain 0 on a reasoning-capable model with the disable toggle on)
- The fix does not regress `response_format` handling (ca4a24d stays correct — `response_format` is not on the wire)
- A wire-body regression test (`openai-compat-request-body.test.ts` style) asserts `repetition_penalty: <value>` is on the request body when the setting is set

## Estimated scope

3-5 days depending on chosen path. Path (2) is preferable if the per-id key correction lands first (tracked separately); otherwise path (1) is the standalone option.

## References

- #402 (closed) — original report
- #411 (Batch 6) — force-disable thinking 4-layer fallback, same zod-strip mechanism
- ca4a24d — `response_format` removal
- `src/llm-sdk/openai-compat-sdk-client.ts:343-417` — current `buildProviderOptions`

**Comment by green-dalii** [comment]
Routing through v1.26.x PATCH (item 5 in our ROI table — `~/.claude/projects/-Users-greener-project-obsidian-llm-wiki/memory/project_v1_26_x_patch_scope.md`). The setting currently has zero wire-level effect on the openai-compatible provider.

Note: the `#402` close reply mentioned `#415` — that was a typo. The split-off is `#414` (this issue). Apologies for the noise.

Two repair paths I can see, similar shape to `#403`/`#407`. I'd like your read on which one to ship before opening the PR:

1. **`transformRequestBody` route per-field** — add a wrapper that injects `repetition_penalty` (snake_case) into the request body directly when `this.providerOptionsName === 'openai-compatible'`. Pros: scoped to exactly one field, no SDK coupling. Cons: requires per-backend compat measurement (LM Studio / DeepSeek / Kimi / GLM all have different responses to the field — DocTpoint's #402 measurements suggested dropping rather than routing it for `response_format`; the analogous measurement for `repetition_penalty` is what's missing).

2. **Per-id key correction** — switch the wrapper to emit `{ [this.provider]: openaiOpts }` instead of `{ openaiCompatible: openaiOpts }`, matching the SDK's actual passthrough key (per your DocTpoint #382 comment 2 analysis). Pros: unlocks the Layer-2 path-2 passthrough for ALL extra fields at once, not just `repetition_penalty`. Cons: changes the wrapper contract (item 10 in our scope) — same prerequisite: Layer-3 400-retry must be live (now met via PR #411).

I lean (1) for `repetition_penalty` specifically (smaller blast radius, isolated measurement), and would only consider (2) if we ship a per-call `thinkingPolicy` enum (item 6) at the same time so we have the per-call policy machinery to gate the new key.

Estimated scope: 1 commit, 5-15 lines depending on path, plus the per-backend measurement for (1) or the wrapper-key migration test for (2).

**Comment by DocTpoint** [comment]
Per-backend measurement for `repetition_penalty` — LM Studio, re-run today with a control arm.

**Setup.** LM Studio, `google/gemma-4-12b`, MLX runtime, 4bit, arch `gemma4_unified`. Direct HTTP against `/v1/chat/completions`, no plugin in the path. Greedy decoding (`temperature: 0`): a repetition penalty changes the logits before the argmax, so its effect stays visible without sampling. Same prompt every time, 220 max tokens, compared on difference rather than direction.

| field in the request body | HTTP | output vs. baseline |
|---|---|---|
| baseline, sent twice | 200 | identical — tool control |
| `temperature: 1.4` | 200 | different — tool control |
| `repeat_penalty: 1.8` | 200 | **different**, reproduced twice byte-identical |
| `repetition_penalty: 1.8` | 200 | identical |
| `frequency_penalty: 1.8` | 200 | identical |
| `presence_penalty: 1.8` | 200 | identical |
| an invented field name | 200 | identical |

**A cheaper discriminator than the output diff.** Send a type error and let the server's own schema answer:

```
repeat_penalty:     "abc"  ->  HTTP 400   Field with key llm.prediction.repeatPenalty
                                          does not satisfy the schema: expected number
repetition_penalty: "abc"  ->  HTTP 200
frequency_penalty:  "abc"  ->  HTTP 200
an invented field:  "abc"  ->  HTTP 200
```

This backend does validate the fields it knows, and `repetition_penalty` is not one of them — it is discarded as an unknown key, with a 200 and no warning. That is a one-line check anyone can repeat against their own backend, without needing a stable output to compare.

**What it means for the two paths.** Both of them put the spelling `repetition_penalty` on the wire. On this backend that spelling is inert however it arrives, so neither path on its own can satisfy acceptance criterion 1 (`repetitionPenalty: 1.1` vs `1.0` measurably different on LM Studio / gemma-4-12b). Criterion 4, the wire-body regression test, would turn green while the user setting still does nothing here. The comment above `buildProviderOptions` already reads the situation this way for llama.cpp b9205; this is the same result on a second backend, this time with an artefact that can be re-run.

**Our read, since you asked for one.** We also lean towards path 1, with one addition: the transform needs to choose the spelling per backend, not just inject a fixed one — `repeat_penalty` for LM Studio and llama.cpp, `repetition_penalty` for vLLM and OpenRouter. Without that, path 1 ships the field to the backends that already ignore it. Path 2 has the same gap and a wider blast radius. One point in favour of doing it in the body transform: the pre-v1.23.0 `unsupportedFields` blocklist learned from 400 responses, and on this backend an unread field returns 200 — so a revived blocklist would never learn about it.

We have no measurement for DeepSeek, Kimi, GLM, Ollama or vLLM, so we cannot say how wide the dialect split really is.

Method note: we did not use a fixed seed. On this backend the `seed` field was not honoured in our tests, which makes a seeded comparison noise; that is a separate report.

**Comment by green-dalii** [comment]
Fixed in v1.26.3: https://github.com/green-dalii/obsidian-llm-wiki/releases/tag/1.26.3

**Per-provider dialect dispatch** in `OpenAICompatSdkClient.buildProviderOptions` (PR #453): `lmstudio` / `ollama` → wire `repeat_penalty` (no `-ion`); `kimi` / `openrouter` / `custom` → wire `repetition_penalty` (snake_case, OpenAI-spec); `deepseek` / `gemini` / `minimax` / `glm` / `bedrock-openai` / unknown → field dropped silently. The Anthropic client now drops the field entirely instead of placing an unrecognized key on the wire — matches the 10-locale i18n text *"cloud providers will silently ignore it"*. The return key flips from `{ openaiCompatible: openaiOpts }` to `{ [this.provider]: openaiOpts }` so the AI SDK's per-id-key passthrough at `@ai-sdk/openai-compatible@2.0.62/dist/index.mjs:525-540` delivers the field unchanged.

A one-line `[REPETITION-PENALTY-EMIT]` console.debug includes `baseURL=` so users with developer-mode debugging can verify the wire contract on their backend.

**Known limitation** (tracked as #451 for v1.27.0): `wrapWithAdvancedSettings` (`src/llm-client-wrapper.ts`) uses `Object.create(client)` to inherit `createMessageStream` without settings injection, so `repetitionPenalty` (and all other settings) is silently dropped on the stream path (Query Wiki, streaming UI). The fix in this PR lands only on the non-stream path. Stream-path users will not see a change this release.

Closes #414.
